text/0000-tracked-built-ins.md

```diff
@@ -0,0 +1,270 @@
+
+class Foo {
+  @tracked value = 123;
+
+  obj = new TrackedObject<{[string: number]}>({ foo: 123 });
+  arr = new TrackedArray<string, number>(['foo', 123]);
+  map = new TrackedMap<string, number>(['foo', 123]);
+  set = new TrackedSet<string, number>(['foo', 123]);
+  weakMap = new TrackedWeakMap<{[object: number]}>([[{}, 456]]);
+  weakSet = new TrackedWeakSet<string, number>(['foo', 123]);
+}
+```
+
+or using `tracked()` decorator:
+
+```ts
+import { tracked } from 'tracked-built-ins';
+
+class Foo {
+  @tracked value = 123;
+
+  obj = tracked<{[string: number]}>({ foo: 123 });
+  arr = tracked<string, number>(['foo', 123]);
+  map = tracked<string, number>(new Map(['foo', 123]));
+  set = tracked<string, number>(new Set(['foo', 123]));
+  weakMap = tracked<{[string: number]}>(new WeakMap([[{}, 456]]));
+  weakSet = tracked<string, number>(new WeakSet(['foo', 123]));
+}
+```
+
+## Detailed design
+
+The necessary changes to `ember-cli` are relatively small since we only need
+to add the dependency to the `app` blueprint, and the `addon` blueprint will
+inherit it automatically.
+
+This has the advantage (over including it as an implicit dependency), that
+apps and addons that don't want to use it for some reason can opt out by
+removing the dependency from their `package.json` file.
+
+## How we teach this
+
+The following guide should be added to the Autotracking In-Depth guide in the official guides, and the
+sections on [POJOs][guides-pojos] and [Array][guides-arrays] should be removed.
+
+### Tracked data structures
+
+Generally, you should try to create classes with their tracked properties enumerated
+and decorated with `@tracked`, instead of relying on dynamically created objects.
+In some cases however, there could be a prohibitive number of possible properties,
+or there could be no way to know them in advance.
+In this case, you should use tracked versions of JavaScript's built-in
+Plain Old JavaScript Object (POJO), array or keyed collection (Maps, Sets, WeakMaps, WeakSets)
+provided by `tracked-built-ins` package.
+
+```js
+import {
+  TrackedObject,
+  TrackedArray,
+  TrackedMap,
+  TrackedSet,
+  TrackedWeakMap,
+  TrackedWeakSet,
+} from 'tracked-built-ins';
+```
+
+For example, we could make a scoreboard component that keeps score for an
+arbitrary number of players, and keep track of the score for each player using
+tracked version of JavaScript [Map][global-objects-map].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedMap } from 'tracked-built-ins';
+
+export default class Scoreboard extends Component {
+  // Mock data for players - this would eventually be replaced with an argument
+  // so players could be passed in
+  players = [
+    { name: 'Zoey' },
+    { name: 'Tomster' },
+  ];
+
+  // Create a map from player -> score, with each player's score starting at 0
+  scores = new TrackedMap(this.players.map(p => [p, 0]));
```

Comment:
We should remove this comment, as this code sample now suggests a pattern we do *not* want to use. (I know this came from the tracked-built-ins README, so no worries!) As written, this suggests a pattern that will actually lead people to bugs if they do as it suggests: the `TrackedMap` will *not* update if they pass in args defining the players. While we could make that a getter, it would then need to be `@cached`, and that increases the complexity quite a bit. Here, I'm suggesting we just explicitly make the players a hard-coded value and show creating a per-instance `scores` which uses the players as its inputs instead.

```suggestion
const PLAYERS = [
  { name: 'Zoey' },
  { name: 'Tomster' },
];

export default class Scoreboard extends Component {
  // Create a map from player -> score, with each player's score starting at 0
  scores = new TrackedMap(PLAYERS.map(p => [p, 0]));
```

Reply:
totally agree! Class fields should be used cautiously as they are not reactive